### ISSUE TITLE
Sample unconfirmed Training Code - I am trying to train

### ISSUE THREAD

**Issue by karayakar** [nari-labs/dia#48]
I have made some changes, training started, lets see waiting one epoch to finish..

```
import os
import torch
import torchaudio
import pandas as pd
from torch.utils.data import Dataset, DataLoader
from tqdm import tqdm

from config import DiaConfig
from layers import DiaModel
from audio import  build_delay_indices, apply_audio_delay
import dac

device = "cuda" if torch.cuda.is_available() else "cpu"

class DiaDataset(Dataset):
    def __init__(self, csv_path, audio_root, config: DiaConfig):
        self.df = pd.read_csv(csv_path, sep='\s*\|\s*', engine='python', names=['audio', 'text'])
        self.audio_root = audio_root
        self.config = config

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        row = self.df.iloc[idx]
        text = row['text']
        audio_path = os.path.join(self.audio_root, row['audio'])
        waveform, sr = torchaudio.load(audio_path)
        if sr != 44100:
            waveform = torchaudio.functional.resample(waveform, sr, 44100)
        return text, waveform


def collate_fn(batch, config, dac_model, device):
    from torch.nn.functional import pad

    texts, audios = zip(*batch)
    max_text_len = config.data.text_length
    text_pad = config.data.text_pad_value
    audio_bos = config.data.audio_bos_value
    audio_pad = config.data.audio_pad_value
    max_audio_len = config.data.audio_length

    text_ids = []
    for txt in texts:
        byte_ids = list(txt.encode("utf-8"))[:max_text_len]
        padded = byte_ids + [text_pad] * (max_text_len - len(byte_ids))
        text_ids.append(torch.tensor(padded, dtype=torch.long))

    text_tensor = torch.stack(text_ids)
    text_positions = torch.arange(max_text_len).unsqueeze(0).expand(text_tensor.size(0), -1)
    text_mask = (text_tensor != text_pad)

    max_len = max(a.shape[1] for a in audios)
    padded_audios = [
        pad(a, (0, max_len - a.shape[1])) if a.shape[1] < max_len else a[:, :max_len]
        for a in audios
    ]
    audio_tensor = torch.stack(padded_audios).to(device)

    #codebooks = audio_to_codebook(dac_model, audio_tensor, config.data).to(device)  # [B, T, C] with delay already applied
    with torch.no_grad():
        _, encoded, _, _, _ = dac_model.encode(dac_model.preprocess(audio_tensor, 44100), n_quantizers=None)
        encoded = encoded.transpose(1, 2)  # [B, T, C]

        B, T, C = encoded.shape
        delay_pattern = config.data.delay_pattern
        t_idx, indices = build_delay_indices(B, T, C, delay_pattern)

        codebooks = apply_audio_delay(
            audio_BxTxC=encoded,
            pad_value=config.data.audio_pad_value,
            bos_value=config.data.audio_bos_value,
            precomp=(t_idx, indices),
        )
         
    codebooks = codebooks[:, :max_audio_len, :]

    B, T, C = codebooks.shape
    tgt = torch.full((B, T + 1, C), audio_pad, dtype=torch.long, device=device)
    tgt[:, 0] = audio_bos
    tgt[:, 1:T+1] = codebooks

    tgt_positions = torch.arange(T + 1).unsqueeze(0).expand(B, -1)

    return {
        "src_tokens": text_tensor.to(device),
        "src_positions": text_positions.to(device),
        "src_mask": text_mask.to(device),
        "tgt_tokens": tgt,
        "tgt_positions": tgt_positions.to(device)
    }


def train(model, config, dac_model, csv_path, audio_dir, device="cuda"):
    dataset = DiaDataset(csv_path, audio_dir, config)
    loader = DataLoader(dataset, batch_size=2, shuffle=True,
                        collate_fn=lambda b: collate_fn(b, config, dac_model, device))

    optimizer = torch.optim.AdamW(model.parameters(), lr=3e-4)
    model.to(device).train()

    os.makedirs("checkpoints", exist_ok=True)

    for epoch in range(5):
        pbar = tqdm(loader, desc=f"Epoch {epoch+1}")
        total_loss = 0
        for step, batch in enumerate(pbar):
            src = batch["src_tokens"]
            src_pos = batch["src_positions"]
            tgt = batch["tgt_tokens"]
            tgt_pos = batch["tgt_positions"]

            enc_mask = (src != config.data.text_pad_value).unsqueeze(1).unsqueeze(2)
            dec_mask = (tgt != config.data.audio_pad_value).any(dim=-1).unsqueeze(1).unsqueeze(2)
            cross_mask = enc_mask.expand(dec_mask.size(0), 1, dec_mask.size(2), src.size(1))

            logits = model(
                src_BxS=src,
                tgt_BxTxC=tgt,
                src_positions=src_pos,
                tgt_positions=tgt_pos,
                enc_self_attn_mask=enc_mask,
                dec_self_attn_mask=dec_mask,
                dec_cross_attn_mask=cross_mask,
                enable_dropout=True,
            )

            logits = logits[:, :-1]
            targets = tgt[:, 1:]
            loss = 0
            for c in range(config.data.channels):
                loss += torch.nn.functional.cross_entropy(
                    logits[:, :, c, :].reshape(-1, logits.size(-1)),
                    targets[:, :, c].reshape(-1),
                    ignore_index=config.data.audio_pad_value
                )

            optimizer.zero_grad()
            loss.backward()
            torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
            optimizer.step()

            total_loss += loss.item()
            pbar.set_postfix({"loss": loss.item()})

        checkpoint_path = f"checkpoints/model-epoch{epoch+1}.pth"
        torch.save(model.state_dict(), checkpoint_path)
        print(f"\n✅ Saved checkpoint: {checkpoint_path}")


if __name__ == "__main__":
    config = DiaConfig.load("config.json")
    dac_model = dac.DAC.load(dac.utils.download()).cuda()
    model = DiaModel(config)

     
    train(
        model=model,
        config=config,
        dac_model=dac_model,
        csv_path="....meta_100.csv",
        audio_dir="..audios",
        device="cuda" if torch.cuda.is_available() else "cpu"
    )



```

**Comment by satheeshKOLA532** [comment]
Hello @karayakar , how's the Performance after one epoch?

**Comment by karayakar** [comment]
> Hello [@karayakar](https://github.com/karayakar) , how's the Performance after one epoch?

Didn't finish 1 full epoch.  got some noise.
https://vocaroo.com/1eR0DGbhaaPJ

**Comment by saiful9379** [comment]
@karayakar thank for sharing training code, i am using your training code but this code raise issue like that,

**Comment by ruby11dog** [comment]
I noticed you didn't use CFG in logits during training. Have you considered incorporating it?

**Comment by karayakar** [comment]
https://github.com/anan235/dia-multilingual

**Comment by ruby11dog** [comment]
> https://github.com/anan235/dia-multilingual

I don't see the CFG  strategy being applied in this project's training either.👀

**Comment by karayakar** [comment]
Whenever I hear or saw Style2 TTS it is a messy and not working projects. WTF....
That is why  I don't work on Style2TTS. Style2 TTS is just a research project nothing works, spend 1 month to train with no luck...

Try OprheusTTS or OuteTTS much better easy to train... Works as expected. I am not going to spend my time on bullshit half projects. 
If you see Style2TTS go far away...

**Comment by stlohrey** [comment]
@karayakar 

I made a working finetuning script based on your code. I finetuned it on German HUI Corpus:
[opendata-iisys-hui](https://huggingface.co/datasets/Paradoxia/opendata-iisys-hui)

Sample output from the finetuned model after 30k steps:
https://vocaroo.com/12VN4SLG9WEk

you can find the code here:

https://github.com/stlohrey/dia-finetuning

**Comment by stlohrey** [comment]
> [@stlohrey](https://github.com/stlohrey) Das klingt gut! Surprising good for such a small dataset.
> 
> May I ask you what GPU setup and hyperparameters did you use?
@C00reNUT I trained on a single 3090. You can check the hyperparameters in finetune.py. I had to train with batchsize=1 and reduce text length and audio length to avoid oom.

**Comment by stlohrey** [comment]
@C00reNUT I didn't filter the dataset. I just changed the lengths in the config file, and in my collate_fn the inputs are truncated if they are too long.
I would like to see a finetune on a larger, more modern dataset, and also to check out multilingual training :)

**Comment by karayakar** [comment]
Since I don't have any interest on this project. I am going to close the issue. Feel free to continue on other issue. is That ok?
